README.md
# base-cookbook

Base is a application cookbook around base cookbooks and recipes we want to include in our environments.

Some are default as we need them to be on our nodes, some are attribute driven so we can specify if we want to include them.

## Supported Platforms

- Ubuntu

## Attributes

## default.rb

Inlcude the [apt](https://github.com/opscode-cookbooks/apt) cookbook

```ruby
default['base']['include']['apt'] = true
```

Include the [chef_client](https://github.com/opscode-cookbooks/chef-client) cookbook

```ruby
default['base']['include']['chef-client'] = false
```

Include a [simple report handler](http://jtimberman.housepub.org/blog/2011/04/24/a-simple-report-handler/) which outputs updated resources to stdout

```ruby
default['base']['include']['simple_report_handler'] = true
```

Include the [hipchat handler](https://github.com/cwjohnston/chef-hipchat)

```ruby
default['base']['include']['hipchat_handler'] = false
```

Include the [slack handler](https://github.com/iteh/chef-slack_handler)

```ruby
default['base']['include']['slack_handler'] = false
```

Include [chef-sugar](https://github.com/sethvargo/chef-sugar)

```ruby
default['base']['include']['chef-sugar'] = true
```

Include the [ntp](https://github.com/opscode-cookbooks/ntp/) cookbook

```ruby
default['base']['include']['ntp'] = true
```

Include motd

```ruby
default['base']['include']['motd'] = true
default['base']['motd']['operation_tools'] = true
```

Include some operations tools

```ruby
default['base']['include']['operation_tools'] = true
default['base']['operation-tools'] = %w( htop iotop apachetop atop nmon dstat iftop nload )
```

Include [build_essential](https://github.com/opscode-cookbooks/build-essential) cookbook

```ruby
default['base']['include']['build_essential'] = false
```


## Usage

### chef_base::default

Include `base` in your node's `run_list`:

```json
{
  "run_list": [
    "recipe[base::default]"
  ]
}
```

## Contributing

1. Fork the repository on Github
2. Create a named feature branch (i.e. `add-new-recipe`)
3. Write you change
4. Write tests for your change (if applicable)
5. Run the tests, ensuring they all pass
6. Submit a Pull Request

## License and Authors

Author:: Edmund Haselwanter (me@ehaselwanter.com)


Changelog.md
# Changelog

## 0.4.5

- update to finstyle 1.5 and fix files to use this config [view commit](http://github.com///commit/978ea379323bc42e9f0f92dc03312b2621aef38e)
- use current stable kitchen [view commit](http://github.com///commit/ffb179f517fce8511015fe9fc9f40ed64e78f11e)
- line prefix bug [view commit](http://github.com///commit/ba0043e4de9c2ee281edb5f1e3ec06658676ae6d)
- allow for minor versions of chef [view commit](http://github.com///commit/2e4dcfc48fc7256373d461683b405e9c06b2d24b)
- update change log [view commit](http://github.com///commit/4dbc16903a9f2797b0aa09d0cd5ca561a08c9580)
- version bump and add stove extended metadata [view commit](http://github.com///commit/d8808d7a6775fa04908890431bd9ae4b487515fa)

## 0.4.4

- pin chef-handler-updated-resources gem due to bug in gem [view commit](http://github.com///commit/bf4c2478798a2dda3a8653c9eb12879797c7ee2d)
- fix rubocop complains [view commit](http://github.com///commit/9552638c446d7e2255a817e971443b8552375058)

## 0.4.3

- add ubuntu and omnibus updater [view commit](http://github.com///commit/a6b5524ab75763075ec8bdd5117d62e7a3708e9d)
- remove hipchat from default test [view commit](http://github.com///commit/1bf0e14168d56ad157f82fd94437986181db37c2)

## 0.4.2

- fix rubocop issues [view commit](http://github.com///commit/576c6feb469712c860396a13ac322259c24c1710)
- fix renaming from base to il-base issues [view commit](http://github.com///commit/2bc3f25e7d79fe715a8b17b76f221f3a70fa3047)

## 0.4.1

- rename with a prefix so that we can upload it to the supermarket - move to new home [view commit](http://github.com///commit/fb3b2463db91bc902cd1ceffe25d91853132310a)
- update to recent attribute naming [view commit](http://github.com///commit/b382d4cf27969170ca1705aaf447c00cf769a572)
- version bump [view commit](http://github.com///commit/1b1ae29ae2f6de8f1824436b9be522639f274561)
- Use public url [view commit](http://github.com///commit/175caf075df8a6fe0267ccb7cc50b55d477565dc)

## 0.2.2

- add tmux
- add kitchen-fog

spec/spec_helper.rb
require "chefspec"
require "chefspec/berkshelf"

RSpec.configure do |_config|
  # Specify the path for Chef Solo to find cookbooks
  # config.cookbook_path = '/var/cookbooks'

  # Specify the path for Chef Solo to find roles
  # config.role_path = '/var/roles'

  # Specify the Chef log_level (default: :warn)
  # config.log_level = :debug

  # Specify the path to a local JSON file with Ohai data
  # config.path = 'ohai.json'

  # Specify the operating platform to mock Ohai data from
  # config.platform = 'ubuntu'

  # Specify the operating version to mock Ohai data from
  # config.version = '12.04'

  # Use color output for RSpec
  # config.color_enabled = true

  # Use documentation output formatter
  # config.formatter = :documentation
end


spec/default_spec.rb
require "spec_helper"

# Write unit tests with ChefSpec - https://github.com/sethvargo/chefspec#readme
describe "il-base::default" do
  let(:chef_run) do
    ChefSpec::SoloRunner.new do
      # nedded for the chef-client::delete_validation
      Chef::Config[:client_key] = "a dummy client"
    end.converge(described_recipe)
  end

  %w[apt vim tmux].each do |rcp|
    it "should include the #{rcp} recipe" do
      expect(chef_run).to include_recipe rcp
    end
  end

  it "should not include chef-client" do
    expect(chef_run).to_not include_recipe "chef-client::default"
  end
end


test/integration/default/serverspec/spec_helper.rb
# encoding: utf-8
#
# Copyright 2015, Edmund Haselwanter
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

if ENV['STANDALONE_SPEC']

  require 'serverspec'
  require 'pathname'
  require 'net/ssh'
  require 'highline/import'

  set :backend, :ssh

  RSpec.configure do |c|
    if ENV['ASK_SUDO_PASSWORD']
      c.sudo_password = ask('Enter sudo password: ') { |q| q.echo = false }
    else
      c.sudo_password = ENV['SUDO_PASSWORD']
    end

    options = {}

    if ENV['ASK_LOGIN_PASSWORD']
      options[:password] = ask("\nEnter login password: ") { |q| q.echo = false }
    else
      options[:password] = ENV['LOGIN_PASSWORD']
    end

    if ENV['ASK_LOGIN_USERNAME']
      options[:user] = ask("\nEnter login username: ") { |q| q.echo = false }
    else
      options[:user] = ENV['LOGIN_USERNAME'] || ENV['user'] || Etc.getlogin
    end

    if options[:user].nil?
      puts 'specify login user env LOGIN_USERNAME= or user='
      exit 1
    end

    c.host = ENV['TARGET_HOST']
    c.ssh_options = options.merge(Net::SSH::Config.for(c.host))
  end

else
  require 'serverspec'

  set :backend, :exec
end


recipes/simple_report_handler.rb
#
# Cookbook Name:: base
# Recipe:: simple_report_handler
#
# Copyright (C) 2014 Edmund Haselwanter
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

chef_gem "chef-handler-updated-resources"
require 'chef/handler/updated_resources'

include_recipe "chef_handler"

chef_handler 'SimpleReport::UpdatedResources' do
  source 'chef/handler/updated_resources'
  action :nothing
end.run_action(:enable)


attributes/default.rb
#
# Cookbook Name:: chef_base
# Attribute:: default
#
# Copyright (C) 2015 Edmund Haselwanter
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

# Inlcude the [apt](https://github.com/opscode-cookbooks/apt) cookbook
default['base']['include']['apt'] = true

# Inlcude the [ubuntu](https://supermarket.chef.io/cookbooks/ubuntu) cookbook
default['base']['include']['ubuntu'] = true

# Include the [chef_client](https://github.com/opscode-cookbooks/chef-client) cookbook
default['base']['include']['chef-client'] = false

# Include the [omnibus_updater](https://supermarket.chef.io/cookbooks/omnibus_updater)
default['base']['include']['omnibus-updater'] = true

# Include a [simple report handler](http://jtimberman.housepub.org/blog/2011/04/24/a-simple-report-handler/) which outputs updated resources to stdout
default['base']['include']['simple_report_handler'] = true

# Include the [hipchat handler](https://github.com/cwjohnston/chef-hipchat)
default['base']['include']['hipchat_handler'] = false

# Include the [slack handler](https://github.com/iteh/chef-slack_handler)
default['base']['include']['slack_handler'] = false

# Include [chef-sugar](https://github.com/sethvargo/chef-sugar)
default['base']['include']['chef-sugar'] = true

# Include the [ntp](https://github.com/opscode-cookbooks/ntp/) cookbook
default['base']['include']['ntp'] = true

# Include motd
default['base']['include']['motd'] = true
default['base']['motd']['operation_tools'] = true

# Include some operations tools
default['base']['include']['operation_tools'] = true
default['base']['operation-tools'] = %w[htop iotop apachetop atop nmon dstat iftop nload]

# Include [build_essential](https://github.com/opscode-cookbooks/build-essential) cookbook
default['base']['include']['build_essential'] = false
